nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestGetOrCreateDocument.java

```diff
@@ -0,0 +1,229 @@
+                                        Thread.currentThread().interrupt(); // restore interrupt
+                                        throw new NuxeoException(e);
+                                    }
+                                }
+                                return s.createDocument(docModel);
+                            });
+                } catch (ConcurrentUpdateException e) {
+                    if (latch != null) {
+                        latch.countDown();
+                    }
+                    throw e;
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt(); // restore interrupt
+                    if (latch != null) {
+                        latch.countDown();
+                    }
+                    throw new NuxeoException(e);
+                }
+            });
+        }
+    }
+
+    protected String computeKey(CoreSession session, DocumentModel docModel) {
+        String repositoryName = docModel.getRepositoryName();
+        String parentId = session.getDocument(docModel.getParentRef()).getId();
+        String name = docModel.getName();
+        return repositoryName + "-" + parentId + "-" + name;
+    }
+
+    @Test
+    public void testCreateDocument() throws Exception {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+        assertFalse(session.exists(doc.getRef()));
+
+        GetOrCreateDocumentThread t = new GetOrCreateDocumentThread(session, doc);
+        Thread thread = new Thread(t);
+        thread.start();
+        thread.join();
+
+        doc = t.documentModel;
+
+        assertTrue(session.exists(doc.getRef()));
+
+    }
+
+    @Test
+    public void testGetDocument() throws Exception {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+        doc = session.createDocument(doc);
+        session.save();
+
+        assertTrue(session.exists(doc.getRef()));
+
+        GetOrCreateDocumentThread t = new GetOrCreateDocumentThread(session, doc);
+        Thread thread = new Thread(t);
+        thread.start();
+        thread.join();
+        doc = t.documentModel;
+
+        assertTrue(session.exists(doc.getRef()));
+
+    }
+
+    @Test
+    public void testLockIsSetWhenCreatingDocument() throws Exception {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+
+        KeyValueStore kvStore = Framework.getService(KeyValueService.class).getKeyValueStore(DOCUMENT_LOCK);
+        String key = computeKey(session, doc);
+
+        assertNull(kvStore.get(key));
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        GetOrCreateDocumentThread t = new GetOrCreateDocumentThread(session, doc, latch);
+        Thread thread1 = new Thread(t);
+
+        MutableObject<byte[]> me = new MutableObject<>();
+        Thread thread2 = new Thread(() -> TransactionHelper.runInTransaction(() -> {
```

Comment:
Still discussing the algo in slack, but testing must be more aggressive you can not rely on 2 threads creating 1 documents, make the number of threads and number of document configurable. Default to 10 threads and 50 docs but do test it with 256 threads and 10k docs.

Reply:
Right, but I'm thinking of doing this test with the actual _getOrCreateDocument_ method that I'll add later in the `CoreSession` once we have decided on the right algo for the locking process. Here I just wanted to assert that an exception is thrown if an other thread is trying to access a locked document.